### PR TITLE
Fix AST compilation in netcoreapp

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
The FSCS untyped AST compilation facility is currently broken in .NET Core since the implementation is hardcoding the target profile to mscorlib.

This PR:
* Exposes a `PrimaryAssembly` parameter in the AST Compile* functions.
* Adds a couple of smoke tests to verify correctness.
* Add missing compilation phase to ast compilation pipeline.

### CODE REVIEW

File: tests/fsharp/Compiler/CompilerAssert.fs
Comment (eiriktsarpalis) [inline tests/fsharp/Compiler/CompilerAssert.fs]: I notice elsewhere in this file that calls to the typechecker are being synchronized. AFAIK these api's should be threadsafe though?

File: src/fsharp/service/service.fsi
Comment (eiriktsarpalis) [inline src/fsharp/service/service.fsi]: To Consider:
* Should we expose `PrimaryAssembly` or use a different parameter? `PrimaryAssembly` is awkwardly located in the `FSharp.Compiler.AbstractIL` which has to be opened as well.
* What should the default value be? Currently it just uses the `TcConfigB` default which is mscorlib. However I'm wondering if using a default based on the current runtime might be more sensible.

Comment (cartermp) [reply src/fsharp/service/service.fsi]: Yeah the default should be determined by TFM.

Generally I wouldn't like to expose this info since it feels very much like an implementation detail. I removed some IVTs that snagged items from absil and although this is a better approach it'd be nice if it weren't necessary.

Also idk what a PrimaryAssembly is

Comment (eiriktsarpalis) [reply src/fsharp/service/service.fsi]: > Also idk what a PrimaryAssembly is

It corresponds to the `--targetprofile` argument accepted by the compiler. It has three possible values: `mscorlib`, `netstandard` or `netcore`. The TcConfig default is `mscorlib` regardless of tfm or current runtime.

Comment (eiriktsarpalis) [reply src/fsharp/service/service.fsi]: So to answer your questions, I think it should be exposed, since the cli does the same. The question however is whether we should use a special type for the FsharpChecker API and map to PrimaryAssembly under the hood.

Comment (cartermp) [reply src/fsharp/service/service.fsi]: Gotcha. This seems fine then, since it's a CLI option.

Comment (baronfel) [reply src/fsharp/service/service.fsi]: @cartermp this primaryassembly option is actually the thing o fixed that fixed FSAC FSI typechecking. It wasn't obvious to me because we use ProjectOptions, where this is hidden behind the --targetprofile flag. So I'm 💯 in support of making this API a nice strong DU.

**Comment by TIHan** [comment]
After talking with @terrajobst , having the notion of a `PrimaryAssembly` is fine but we should not be explicit about what `PrimaryAssembly` we want. Instead, it should decide what the `PrimaryAssembly` is based on the given references and there are rules about it, which are described here for C#: http://sourceroslyn.io/#Microsoft.CodeAnalysis/ReferenceManager/CommonReferenceManager.Binding.cs,536628815dfdb8a2

**Comment by eiriktsarpalis** [comment]
> IMHO, we should not expose PrimaryAssembly. Instead, we should be thinking about trying to remove PrimaryAssembly altogether and getting the compiler to be framework agnostic.

That's fair, however the scope of this PR is to bring the ast compiler on par with the cli arguments, as they currently stand.

**Comment by eiriktsarpalis** [comment]
Rebased on top of latest changes

File: src/fsharp/service/service.fsi
Comment (TIHan) [reply src/fsharp/service/service.fsi]: I've been thinking about this.

It turns out we won't be able to remove the `targetprofile` command line option from Fsc as existing calls to the command line compiler rely on it.

However, exposing the notion of a `targetprofile` as part of an API to the compiler is something I would like to avoid; the core compiler itself really should not care about the profile (even though it does today, but Fsc (CLI compiler) is really the only one that should care). If the reasoning is, the "CLI" does it, then everything that the "CLI" allows should be exposed in some way. Perhaps the APIs from Compile/CompileDynamicAssembly that takes ASTs should also take in CLI args and we simply just ignore any source files listed or see if the source files listed match the ASTs' file name.

Comment (eiriktsarpalis) [reply src/fsharp/service/service.fsi]: > Perhaps the APIs from Compile/CompileDynamicAssembly that takes ASTs should also take in CLI args

I'm assuming you're approaching this from the point of view of breaking changes? I was thinking the same, however for better or for worse that particular subset of the api is accepting individual arguments for things like `--no-framework`, so requiring that suddenly become part of an argv is also a breaking change.

> However, exposing the notion of a targetprofile as part of an API to the compiler is something I would like to avoid

I see your point but for all intents and purposes it is exposed. The goal of this PR is to fix something that is currently not working and missing acceptance tests for it. My goal here is to be unblocked for QuotationCompiler. Should this change in the future, changing the signature should not be an issue, provided the tests still pass.

**Comment by eiriktsarpalis** [comment]
Following conversations with @TIHan, I've changed the api surface so that `PrimaryAssembly` is no longer exposed. Until #8043 is merged, it will determine that value based on the current runtime.

**Comment by cartermp** [comment]
@eiriktsarpalis is this work dependent on #8043?

**Comment by eiriktsarpalis** [comment]
@cartermp not exactly, but we decided it would be less work if we let #8043 merge first.

**Comment by cartermp** [comment]
Makes sense.

**Comment by eiriktsarpalis** [comment]
Are you planning on merging #8043 for 5.0? If not I think I should just go ahead and merge this, since it fixes a broken component of the compiler.

**Comment by cartermp** [comment]
I think that's @TIHan's call, but conservatively I'm going to estimate that it won't get in for F# 5.

**Comment by eiriktsarpalis** [comment]
I've rebased on top of latest master.

**Comment by terrajobst** [comment]
> After talking with @terrajobst , having the notion of a `PrimaryAssembly` is fine but we should not be explicit about what `PrimaryAssembly` we want. Instead, it should decide what the `PrimaryAssembly` is based on the given references and there are rules about it, which are described here for C#: http://sourceroslyn.io/#Microsoft.CodeAnalysis/ReferenceManager/CommonReferenceManager.Binding.cs,536628815dfdb8a2

To expand on this: you should chat with @jaredpar and how Roslyn solved this. AFAIK the core assembly is determined by picking the assembly that in turn has no references. If checking that is expensive, you could try a fast pass by checking for `mscorlib` and `System.Runtime` first and see whether they have references and if they do, probe all references.

**Comment by jaredpar** [comment]
> AFAIK the core assembly is determined by picking the assembly that in turn has no references. 

The core assembly has no references **and** contains `System.Object`. There are completely valid cases where assemblies have no references but are not in fact the core assembly.

> If checking that is expensive, you could try a fast pass by checking for mscorlib and System.Runtime first and see whether they have references and if they do, probe all references

FWIW: This check hasn't ever come up as a perf issue for us

**Review by KevinRansom — APPROVED** [review]

**Comment by KevinRansom** [comment]
Thinks for this @eiriktsarpalis

**Review by cartermp — APPROVED** [review]

**Pull request merged**
